data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/JacksonEvent.java

```diff
@@ -444,7 +444,9 @@ private static boolean isValidKey(final String key) {
                     || c == '-'
                     || c == '_'
                     || c == '@'
-                    || c == '/')) {
+                    || c == '/'
+                    || c == '['
+                    || c == ']')) {
```

Comment:
What is the use case for making brackets valid as part of keys?

Reply:
Yeah. This is not correct. We should not be taking Adam's suggestion in the issue literally.

Reply:
> What is the use case for making brackets valid as part of keys?

When flattening arrays in json, it will result in brackets in keys like `list1[0].list2[0].name`. Is there a concern allowlisting the brackets?

Reply:
@oeyh its not about whether they are valid or not but more about usage of it. Is it necessary to do this way? are there other ways of doing it?

Reply:
At the minimum you will have to add support for [ ] in the expressions. Also, does it impact regex matching given that [ and ] are special characters in regex and so on.

Reply:
> Is it necessary to do this way? are there other ways of doing it?

There are other alternatives to represent list indices like `(0)`, `{0}`, `#0`, `/0`, and we can add support for those formats as options. But using square brackets is still the most common and intuitive way and we should support it for our users.

> At the minimum you will have to add support for [ ] in the expressions. Also, does it impact regex matching given that [ and ] are special characters in regex and so on.

Good points! I can add support for [] in expressions. As for regex, I believe in most cases, the field keys are used as the input string to match with a regex pattern, instead of being part of the regex pattern, so we should be good there.

Reply:
Discussed offline and agreed on allowlisting `[]` in event keys.